### ISSUE TITLE
Testing archi-modelrepository-plugin with Archi 4.0.3-beta1

### ISSUE THREAD

**Issue by cushyno** [archi-contribs/archi-modelrepository-plugin#45]
Some great progress being made with the collaboration plugin and speed improvements in Archi.  Thank you.

Archi 4.0.3-beta1 exhibits a substantial speed improvement over 4.0.1.  I attempted to use this plugin with Archi 4.0.3-beta1, but the application stops responding.

I appreciate the nature of working with beta releases but wished to provide some feedback from my testing.  I hope this helps inform further development work.

OS: Windows 7
Java: java version "1.8.0_31", Java(TM) SE Runtime Environment (build 1.8.0_31-b13)

Tests performed:

1. Archi 4.0.3-beta1 with plugin version 0.2.6: Archi hangs after every attempted collaboration function.  Task Manager  shows increased CPU and Memory usage, then Archi stops responding.
2. Archi 4.0.1 with plugin version 0.2.1:  Added model to collaboration workspace, committed changes, published successfully to GitHub.
3. Archi 4.0.3-beta1 with plugin version 0.2.6: Successfully opens model previously saved with v4.0.1 and shows collaboration history previously saved at test 2.  Archi 4.0.3-beta1 hangs when performing any other collaboration functions. Application not responding similar to test 1.

Checking the working copy status at test 1, git shows the repository state like so:

```
git status
On branch master

Initial commit

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        model/

nothing added to commit but untracked files present (use "git add" to track)

```

Chris Usher

**Comment by Phillipus** [comment]
Hi Chris, thanks for the feedback.

To pinpoint the problem, we'd need more info about the repository (where is it? can you connect via other means? proxy?). It may be timing out simple because it cannot connect.

Phil

**Comment by Phillipus** [comment]
When you say "Archi hangs after every attempted collaboration function" does it do this on a Commit action?

**Comment by jbsarrodie** [comment]
>So you think the issue is related to proxy settings?

My understanding is that the proxy settings are ok.

I really wonder if that would not be linked to the latest changes that export to Grafico for each save. From my understanding this is a **big** model. Can @cushyno confirm ?

BTW, this should be easy to test: just do a small change on the model and save it without using any plugin button or action (ie. just do a "File > Save" or CTRL-S if on Windows). If the issue is linked to the Grafico export this should happen even in this case.

**Comment by cushyno** [comment]
I don't think the proxy setting are a problem.  It works in earlier version.

If I open an existing model and choose "Add Local Model to Workspace".  I can see in Task Manager that CPU usage rises and read/write activity is high.  Then for Archi I see CPU activity and read/write reduces to 0.   Archi has at that point written all files into the "model_repository" folder, (.git, model and images folders), so the GRAPHICO behaviour is working and something is creating the .git folder so jgit library looks to be partially working.  However, the application looks like it is still busy, egg-timer still showing and dialog still displaying but the application title shows "(not responding)".

I kill Archi, restart then try another feature, such as "Commit Changes", with same response - Archi goes busy then stops responding.

It could be a timeout issue, but I don't get that in Archi 4.0.1 + plugin v0.2.1 which works fine (but very slow).
It could be that Archi doesn't recognise that jgit has finished reading/writing files and is still waiting for a response?

**Comment by Phillipus** [comment]
> I really wonder if that would not be linked to the latest changes that export to Grafico for each save.

It doesn't do that anymore.

**Comment by cushyno** [comment]
@jbsarrodie You are correct this is a large model.
Just tested the small change + "File->save".  There is no issue there.  The change was written instantly.

**Comment by jbsarrodie** [comment]
>It doesn't do that anymore.

My bad. I'm still using an older version on the PC I use now :-(

**Comment by cushyno** [comment]
I shall test with other versions of Plugin.  I hadn't done that so far, but I shall try working backwards...

**Comment by cushyno** [comment]
With plugin v0.2.5:
- "Commit" success!
- "Publish" shows an error stating: .\model\strategy\folder.xml (Access is denied).  The model folder does not exist, images folder still exists.  Publish failed.

**Comment by cushyno** [comment]
My apologies, I didn't know about the log.  That would have helped earlier.  There are many hundreds of OutOfMemory messages:


```
!ENTRY org.eclipse.core.jobs 4 2 2017-08-21 12:44:45.536

!MESSAGE Unhandled error

!STACK 0

java.lang.OutOfMemoryError: unable to create new native thread

	at java.lang.Thread.start0(Native Method)

	at java.lang.Thread.start(Unknown Source)

	at org.eclipse.core.internal.jobs.WorkerPool.jobQueued(WorkerPool.java:148)

	at org.eclipse.core.internal.jobs.WorkerPool.startJob(WorkerPool.java:244)

	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:52)

```

No errors since using v0.2.5!

**Comment by Phillipus** [comment]
Using 0.2.6 does this memory leak occur straight away? After running the app for a long time?

**Comment by Phillipus** [comment]
Does it do this for all models? Have you tried with smaller models?

Theory = very large model + exporting to grafico using Jobs = memory loss.

**Comment by jbsarrodie** [comment]
>Theory: very large model + exporting to grafico using Jobs = memory loss.

That's my theory too. If confirmed this is easy to overcome as we can easily set a limit on conccurent jobs in a job group.

**Comment by Phillipus** [comment]
How much system memory does the machine have?

I would test:

- on another machine
- using different models of different sizes

Aim is to pinpoint what causes memory loss.

**Comment by cushyno** [comment]
More testing with Archi 4.0.3-beta1 and plugin v0.2.5.

After deleting local model_repository and adding the remote model from Github (using Collaboration "+" icon), I can successfully save changes, commit, publish and refresh.  It is still slow, but that is not too much of a problem because committing changes is not as frequent as making a "file->save".

This is a slow and aging PC but with 3GB ram.  I shall test on a newer PC when I get chance later this week.

@Phillipus  and @jbsarrodie, thank you both for your help.  The speed improvements of 4.0.3 coupled with a shared repository will make a big difference to Archi.  Good work!

**Comment by Phillipus** [comment]
Can we keep this issue open? We need to solve the issue with 0.2.6.

**Comment by Phillipus** [comment]
Could you try this version? Please unzip and copy the jar file into the plugins folder of Archi.

[org.archicontribs.modelrepository_0.2.7.201708211630.zip](https://github.com/archi-contribs/archi-modelrepository-plugin/files/1239278/org.archicontribs.modelrepository_0.2.7.201708211630.zip)

**Comment by cushyno** [comment]
4.0.3-beta1 with plugin v0.2.7

Application opens a little faster.
Commit is faster.
Publish takes about 3 seconds where before it was taking 20+.  Though this may be affected by our network bandwidth at this time of day, it is a big improvement.

I'm going to try this plugin from scratch with a new repository, including the initial "Add Local model to Workspace".  I'll let you know how the full cycle works as soon as I have run a few tests.

Thanks
Chris.

**Comment by cushyno** [comment]
Here are some rough stats:

- 3s to open my model (Wow, that's fast!)
- 34s for initial commit - Add local model to workspace.
- 9s to commit small change (this will be writing out the whole model like GRAPHICO?  9s is reasonable performance for the model size).
- 37s for initial publish to remote repository
- 8s to commit the next small change
- 11s to publish to remote repository

So initially adding the model takes time because git has to set up a new local working copy.
First publish to remote repository on github takes time because it is transferring all the data to remote  site.
Apart from these two transactions performance seems good.

No memory errors at all.  Original issue has been resolved.
Thank you.

**Comment by Phillipus** [comment]
Thanks for testing. I've released this as the 0.2.7 build now.

**Comment by Phillipus** [comment]
Closing.
